src/cluster/mcl.cpp

```diff
@@ -0,0 +1,443 @@
+	// vector<unordered_set<uint32_t>> cr;
+	// float inf = 2.0;
+	// float exp = 2.0;
+	// if(false){
+	// 	Eigen::SparseMatrix<float> tmp(12,12);
+	// 	tmp.setFromTriplets(data.begin(), data.end());
+	// 	printMatrix(&tmp);
+	// 	cr = markov_process(&tmp, inf, exp);
+	// 	printMatrix(&tmp);
+	// }
+	// else{
+	// 	Eigen::MatrixXf tmp = Eigen::MatrixXf::Zero(12,12);
+	// 	for(Eigen::Triplet<float> const & t : data){
+	// 		tmp(t.row(), t.col()) = t.value();
+	// 	}
+	// 	printMatrix(&tmp);
+	// 	cr = markov_process(&tmp, inf, exp);
+	// 	printMatrix(&tmp);
+	// }
+	// cout << "Found "<<cr.size()<<" clusters"<<endl;
+	// for(auto& set: cr){
+	// 	cout<< "set :";
+	// 	for(auto item : set){
+	// 		cout<< " " << item;
+	// 	}
+	// 	cout<<endl;
+	// }
+	// return;
+
+	const size_t seq_count = db->ref_header.sequences;
+	statistics.reset();
+	config.command = Config::blastp;
+	config.no_self_hits = false;
+	config.output_format = { "clus" };
+	config.query_cover = 80;
+	config.subject_cover = 80;
+	config.algo = 0;
+	config.index_mode = 0;
+	config.freq_sd = 0;
+	config.max_alignments = numeric_limits<uint64_t>::max();
+
+	Workflow::Search::Options opt;
+	opt.db = &(*db);
+	opt.self = true;
+	SparseMatrixStream<float> ms(db->ref_header.sequences);
+	opt.consumer = &ms;
+	opt.db_filter = nullptr;
+	Workflow::Search::run(opt);
+
+	task_timer timer;
+	timer.go("Computing independent components");
+	auto componentInfo = ms.getComponents();
+	vector<vector<uint32_t>> indices = get<0>(componentInfo);
+	vector<vector<Eigen::Triplet<float>>> components = get<1>(componentInfo);
+	uint32_t nComponents = count_if(indices.begin(), indices.end(), [](vector<uint32_t> v){ return v.size() > 0;});
+	uint32_t nComponentsLt1 = count_if(indices.begin(), indices.end(), [](vector<uint32_t> v){ return v.size() > 1;});
+	timer.finish();
+	cout << "Found " << nComponentsLt1 <<" ("<<nComponents<< " incl. singletons) disconnected components" << endl;
+
+	timer.go("Clustering components");
+	bool full = false;
+	vector<uint32_t> clustering_result(db->ref_header.sequences, std::numeric_limits<uint32_t>::max());
+	float max_sparsity = 0.0f;
+	float min_sparsity = 1.0f;
+	uint32_t cluster_id = 0;
+	uint32_t nClustersEq1 = 0;
+	uint32_t n_dense_calculations = 0;
+	uint32_t n_sparse_calculations = 0;
+	float inflation = (float) config.cluster_mcl_inflation;
+	float expansion = (float) config.cluster_mcl_expansion;
+
+	if( full ){
+		// TODO: According to the SIAM publication this is not valid, just for debugging
+		Eigen::SparseMatrix<float> m = ms.getMatrix();
+		for(unordered_set<uint32_t> subset : markov_process(&m, inflation, expansion)){
+			for(uint32_t el : subset){
+				clustering_result[el] = cluster_id;
+			}
+			cluster_id++;
+			if(subset.size() == 1) nClustersEq1 ++;
+		}
+	}
+	else {
+		// TODO: check when/if using dense matrices for non-sparse components improves performance
+		for(uint32_t iComponent = 0; iComponent<indices.size(); iComponent++){
```

Comment:
Should this loop be multithreaded?

Reply:
I think that would help quite a bit. I will look into it.

Reply:
I have now multithreaded this part.